File: ios/Sources/Detonator/Element/ScrollViewElement/ScrollViewElement.swift
import UIKit

class ScrollViewElement: Element {
    override public func decodeAttributes(edge: Edge) -> ScrollViewAttributes? {
        return super.decodeAttributes(edge: edge)
    }
    
    override public func createView() -> ScrollView {
        let view = ScrollView()
        
        view.isViewGroup = true
        
        view.contentInsetAdjustmentBehavior = .never
        
        return view
    }
    
    override public func patchView() {
        let view = view as! ScrollView
        
        let attributes = attributes as! ScrollViewAttributes
        let currentAttributes = currentAttributes as! ScrollViewAttributes?
        
        let showsIndicator = attributes.showsIndicator
        let currentShowsIndicator = currentAttributes?.showsIndicator
        
        let patchShowsIndicator = forcePatch || showsIndicator != currentShowsIndicator
        
        if patchShowsIndicator {
            if showsIndicator == true {
                view.showsVerticalScrollIndicator = true
            } else {
                view.showsVerticalScrollIndicator = false
            }
        }
    }
    
    override func patchBorderRadius(
        borderRadius: StyleSize?,
        borderRadiusTopLeft: StyleSize?,
        borderRadiusTopRight: StyleSize?,
        borderRadiusBottomLeft: StyleSize?,
        borderRadiusBottomRight: StyleSize?
    ) {}
    
    class ScrollViewAttributes: Attributes {
        var horizontal: Bool?
        var showsIndicator: Bool?
        
        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            horizontal = try container.decodeIfPresent(Bool.self, forKey: .horizontal)
            showsIndicator = try container.decodeIfPresent(Bool.self, forKey: .showsIndicator)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case horizontal
            case showsIndicator
        }
    }
}

File: android/src/main/java/com/iconshot/detonator/element/scrollviewelement/ScrollViewElement.java
package com.iconshot.detonator.element.scrollviewelement;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.element.Element;
import com.iconshot.detonator.helpers.CompareHelper;
import com.iconshot.detonator.helpers.ContextHelper;
import com.iconshot.detonator.layout.ViewLayout;

public class ScrollViewElement extends Element<CustomScrollView, ScrollViewElement.Attributes> {
    public ScrollViewElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    public Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    @Override
    public CustomScrollView createView() {
        return new CustomScrollView(ContextHelper.context);
    }

    @Override
    protected void patchView() {
        Boolean showsIndicator = attributes.showsIndicator;

        Boolean currentShowsIndicator = currentAttributes != null ? currentAttributes.showsIndicator : null;

        boolean patchShowsIndicator = forcePatch || !CompareHelper.compareObjects(showsIndicator, currentShowsIndicator);

        if (patchShowsIndicator) {
            if (showsIndicator != null && showsIndicator) {
                view.setVerticalScrollBarEnabled(true);
            } else {
                view.setVerticalScrollBarEnabled(false);
            }
        }
    }

    protected void patchBackgroundColor(
            Object backgroundColor,
            Object borderRadius,
            Object borderRadiusTopLeft,
            Object borderRadiusTopRight,
            Object borderRadiusBottomLeft,
            Object borderRadiusBottomRight
    ) {
        super.patchBackgroundColor(
                backgroundColor,
                borderRadius,
                borderRadiusTopLeft,
                borderRadiusTopRight,
                borderRadiusBottomLeft,
                borderRadiusBottomRight
        );

        ViewLayout.LayoutParams layoutParams = (ViewLayout.LayoutParams) view.getLayoutParams();

        layoutParams.onLayoutClosures.put("borderRadius", () -> {
            float[] radii = getRadii(
                    borderRadius,
                    borderRadiusTopLeft,
                    borderRadiusTopRight,
                    borderRadiusBottomLeft,
                    borderRadiusBottomRight
            );

            view.setRadii(radii);
        });
    }

    protected static class Attributes extends Element.Attributes {
        Boolean horizontal;
        Boolean showsIndicator;
    }
}
